bundledatasrc_test.go

```diff
@@ -0,0 +1,214 @@
+	}{
+		{
+			descr: "reading bundle from local yaml file",
+			getSource: func() (BundleDataSource, error) {
+				path := bundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(filepath.Join(path, "bundle.yaml"))
+			},
+		},
+		{
+			descr: "reading bundle from exploded archive folder",
+			getSource: func() (BundleDataSource, error) {
+				path := bundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(path)
+			},
+		},
+		{
+			descr: "reading bundle from archive",
+			getSource: func() (BundleDataSource, error) {
+				path := archiveBundleDirPath(c, "wordpress-multidoc")
+				return LocalBundleDataSource(path)
+			},
+		},
+		{
+			descr: "reading bundle from stream",
+			getSource: func() (BundleDataSource, error) {
+				r := strings.NewReader(multidocBundle)
+				return StreamBundleDataSource(r, "https://example.com")
+			},
+		},
+	}
+
+	for specIndex, spec := range specs {
+		c.Logf("[test %d] %s", specIndex, spec.descr)
+
+		ds, err := spec.getSource()
+		c.Assert(err, gc.IsNil)
+
+		parts := ds.Parts()
+		c.Assert(parts, gc.HasLen, 3)
+		s.assertFieldPresent(c, parts[1], "applications.wordpress.offers.offer1.endpoints")
+		s.assertFieldPresent(c, parts[2], "applications.wordpress.offers.offer1.acl.admin")
+	}
+}
+
+func (s *BundleDataSourceSuite) TestResolveAbsoluteFileInclude(c *gc.C) {
+	target, err := filepath.Abs(filepath.Join(c.MkDir(), "example"))
+	c.Assert(err, gc.IsNil)
+
+	expContent := []byte("example content")
+	c.Assert(ioutil.WriteFile(target, expContent, os.ModePerm), gc.IsNil)
+
+	ds := new(resolvedBundleDataSource)
+
+	got, err := ds.ResolveInclude(target)
+	c.Assert(err, gc.IsNil)
+	c.Assert(got, gc.DeepEquals, expContent)
+}
+
+func (s *BundleDataSourceSuite) TestResolveRelativeFileInclude(c *gc.C) {
+	relTo := c.MkDir()
+	target, err := filepath.Abs(filepath.Join(relTo, "example"))
+	c.Assert(err, gc.IsNil)
+
+	expContent := []byte("example content")
+	c.Assert(ioutil.WriteFile(target, expContent, os.ModePerm), gc.IsNil)
+
+	ds := &resolvedBundleDataSource{
+		basePath: relTo,
+	}
+
+	got, err := ds.ResolveInclude("./example")
+	c.Assert(err, gc.IsNil)
+	c.Assert(got, gc.DeepEquals, expContent)
+}
+
+func (s *BundleDataSourceSuite) TestResolveIncludeErrors(c *gc.C) {
+	tmpDir := c.MkDir()
+	specs := []struct {
+		descr   string
+		incPath string
+		exp     string
+	}{
+		{
+			descr:   "abs path does not exist",
+			incPath: "/some/invalid/path",
+			exp:     `include file "/some/invalid/path" not found`,
+		},
+		{
+			descr:   "rel path does not exist",
+			incPath: "./missing",
+			exp:     `include file "./missing" not found`,
+		},
+		{
+			descr:   "path points to directory",
+			incPath: tmpDir,
+			exp:     fmt.Sprintf("include path %q resolves to a folder", tmpDir),
+		},
+	}
+
+	ds := new(resolvedBundleDataSource)
+	for specIndex, spec := range specs {
+		c.Logf("[test %d] %s", specIndex, spec.descr)
+
+		_, err := ds.ResolveInclude(spec.incPath)
```

Comment:
do you feel this is easier to maintain/add more than just separate tests? This is borderline for me, so maybe it isn't worth changing anything. But you do end up with log messages that are too long, and extra messages so you have a chance to figure out what is going on. But the line you need to edit *isn't* the line that c.Assert is on. (you have to figure out which test spec needs updating)

Reply:
While personally I prefer table-driven tests (wrapped in t.Run when using just the native go test stuff) I do understand that it's cumbersome to work with them using gocheck. Generally, when I am adding code I try to emulate the code/test style that is already present in the files I tough and IIRC charms has a few table-driven tests.